src/storage/src/storage_state.rs

```diff
@@ -92,6 +96,79 @@ pub struct StorageState {
     /// Frontier of sink writes (all subsequent writes will be at times at or
     /// equal to this frontier)
     pub sink_write_frontiers: HashMap<GlobalId, Rc<RefCell<Antichain<Timestamp>>>>,
+    /// See: [SinkHandle]
+    pub sink_handles: HashMap<GlobalId, SinkHandle>,
+}
+
+/// This maintains an additional read hold on the source data for a sink, alongside
+/// the controller's hold and the handle used to read the shard internally.
+/// This is useful because environmentd's hold might expire, and the handle we use
+/// to read advances ahead of what we've successfully committed.
+/// In theory this could be stored alongside the other sink data, but this isn't
+/// intended to be a long term solution; either this should become a "critical" handle
+/// or environmentd will learn to hold its handles across restarts and this won't be
+/// needed.
+pub struct SinkHandle {
+    downgrade_tx: UnboundedSender<Antichain<Timestamp>>,
+    _handle: JoinHandle<()>,
+}
+
+impl SinkHandle {
+    /// A new handle.
+    pub fn new(
+        persist_location: PersistLocation,
+        shard_id: ShardId,
+        persist_clients: Arc<Mutex<PersistClientCache>>,
+    ) -> SinkHandle {
+        let (downgrade_tx, mut rx) = mpsc::unbounded_channel();
+
+        let _handle = mz_ore::task::spawn(|| "Sink handle advancement", async move {
+            let client = persist_clients
+                .lock()
+                .await
+                .open(persist_location)
+                .await
+                .expect("opening persist client");
+
+            let mut read_handle: ReadHandle<SourceData, (), Timestamp, Diff> = client
+                .open_reader(shard_id)
+                .await
+                .expect("opening reader for shard");
+
+            let mut downgrade_to = read_handle.since().clone();
+            'downgrading: loop {
+                tokio::select! {
+                    downgrade = rx.recv() => {
+                        if let Some(downgrade) = downgrade {
+                            downgrade_to = downgrade;
+                        } else {
+                            // The sending end of this channel has been dropped, which means
+                            // we're no longer interested in this handle.
+                            break 'downgrading;
+                        }
+                    }
+                    _ = sleep(Duration::from_secs(1)) => {}
```

Comment:
Id make an `tokio::time::interval` and select the `interval.tick()`

Also, 1 second is far too often! at least 1 minute is probably fine!

Reply:
OOC, why `interval`?

Reply:
Thinking about this more, I think the sleep has better behaviour... if the current select resolves after 950ms, we still want to wait the full second on the next iteration. (ie. it's a timeout and not a timer.)

Fair point on the 1s->1m. I'll change that now!

Reply:
oh its a timeout, interesting, your call!